Check whether qualifiers in request are supported

tomcoldrick-ct opened the pull request:
Brings us in line with the spec by checking whether we support the supplied qualifiers before Fetching, returning an appropriate error with details. In order to do this consistently, I've  made a "Fetcher" interface around the "remoteasset.FetchServer", which allows us to check for qualifier support. This formalises the close relationship of Fetchers and Qualifiers.

Note that this does mean that if a client `Push`es a blob with unsupported qualifiers, we won't be able to get said blob out of the store. I think we should just trust clients to not push stuff with qualifiers that aren't supported for the moment given the lack of data on how these will be used. Perhaps to get around this we could point the Pusher at the top-level fetcher and call` CheckQualifiers` before pushing?

tomcoldrick-ct committed:
Fetch: Add qualifier checking

Allows us to check whether the qualifiers supplied by the request are
supported by any of the fetchers specified. In order to do this, I've
created a "Fetcher" interface extending the "remoteasset.FetchServer"
interface, which has an additional method to check whether a set of
qualifier names are supported, returning any that are not.

This allows us to discover whether the request is supported by the
server, and propagate the details back to the ValidatingFetcher at the
front to raise an error.

I've also implemented a Set type for qualifier names.

tomcoldrick-ct committed:
Fetch: Add tests for validating Qualifier support

tomcoldrick-ct commented:
Drafting as I think it might be more convenient to return an error rather than a `qualifier.Set`, I wanna see if my thinking is correct.

Qinusty commented:
> Drafting as I think it might be more convenient to return an error rather than a `qualifier.Set`, I wanna see if my thinking is correct.

For the use following use case I think qualifier.Set is an appropriate response for detailing the missing qualifiers.

```
  // In the case of unsupported qualifiers, the server *SHOULD* additionally
  // send a [BadRequest][google.rpc.BadRequest] error detail where, for each
  // unsupported qualifier, there is a `FieldViolation` with a `field` of
  // `qualifiers.name` and a `description` of `"{qualifier}" not supported`
  // indicating the name of the unsupported qualifier.
```

tomcoldrick-ct commented:
> For the use following use case I think qualifier.Set is an appropriate response for detailing the missing qualifiers.

Yeah, my  idea is to convert to an error at the lowest level and push back up with the error rather than a set of qualifiers. My reasoning for this is so that a `GRPCFetcher` which just forwards requests can check this more easily by using a `FetchBlob` request populated with the qualifiers in question, without having to convert twice or implementing an additional internal gRPC service.

The name of the method may need reworking to make this clearer, naming things never was my strong suit.

tomcoldrick-ct committed:
Return `error` when checking qualifiers

This will allow for easier propagation through a gRPC proxying fetcher,
as otherwise we'd have to parse the error manually.

tomcoldrick-ct committed:
Move qualifier_container to qualifier package

Initially qualifierContainer was a part of the storage package as it was
(and still is) only used to implement Sort on an array of qualifiers.
Now that there's a package for similar qualifier-based abstractions it
makes sense to move it there.

Renames qualifierContainer to the more informative qualifer.Sorter.
